File: image2ascii.py
import cv2
import numpy as np
from math import floor
from functools import lru_cache
from concurrent.futures import ThreadPoolExecutor
import argparse
from time import monotonic

@lru_cache(maxsize=None)
def getChar(chars: str, color: int) -> str:
    return chars[floor((color / 256) * len(chars))]

def image2ascii(image,
                scalingFactor: float,
                charOffset: int = 0,
                white_bg: bool = False,
                bw_font: bool = False,
                fontface = cv2.FONT_HERSHEY_SIMPLEX,
                fontScale: float = 0.4,
                thickness: int = 1,
                lineType: int = 1,
                textOffset: float = 1.3):

    # Consistent font face
    fontFace = cv2.FONT_HERSHEY_SIMPLEX

    # Character set based on background
    # chars = "@B8W#ZQOUCXokdpmzcvunxrjft/1{]?-+~<i!I;:."[::-1] + " " * charOffset
    chars = "$@#%&MWBQ8RNmD69KOgSG50HE2PZqpdb3XUw4CeAk[]aho{}FzVYx=sunc()?y+T7J*L<>~jtf/1vr^Iil!-;_:`,'."[::-1] + " " * charOffset
    if white_bg:
        chars = chars[::-1]  # Optional: reverse again if needed for white background

    # Resize image to shrink to ASCII scale
    image = cv2.resize(src=image, dsize=None, fx=scalingFactor, fy=scalingFactor)
    height, width, _ = image.shape

    # Convert to grayscale
    grayscale_image = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)

    # Calculate font padding based on character size
    char_size = cv2.getTextSize(".", fontFace, fontScale, thickness)[0]
    fontPadding = int(max(char_size) * textOffset)

    # Create blank ASCII image canvas
    ascii_height, ascii_width = height * fontPadding, width * fontPadding
    ascii_image = np.zeros((ascii_height, ascii_width, 3), np.uint8)

    # Fill background white if needed
    if white_bg:
        ascii_image.fill(255)

    # Define rendering functions
    def process_row_bwfont_whitebg(i):
        for j in range(width):
            color = grayscale_image[i, j]
            cv2.putText(ascii_image, getChar(chars, color),
                        (j * fontPadding, i * fontPadding),
                        fontFace, fontScale, (0, 0, 0), thickness, lineType)

    def process_row_bwfont_blackbg(i):
        for j in range(width):
            color = grayscale_image[i, j]
            cv2.putText(ascii_image, getChar(chars, color),
                        (j * fontPadding, i * fontPadding),
                        fontFace, fontScale, (255, 255, 255), thickness, lineType)

    def process_row_colorfont_blackbg(i):
        for j in range(width):
            color = grayscale_image[i, j]
            pixel = image[i, j].tolist()
            cv2.putText(ascii_image, getChar(chars, color),
                        (j * fontPadding, i * fontPadding),
                        fontFace, fontScale, pixel, thickness, lineType)

    # Use threads for speed
    with ThreadPoolExecutor() as executor:
        if bw_font:
            if white_bg:
                executor.map(process_row_bwfont_whitebg, range(height))
            else:
                executor.map(process_row_bwfont_blackbg, range(height))
        else:
            executor.map(process_row_colorfont_blackbg, range(height))

    return ascii_image


if __name__ == '__main__':


    parser = argparse.ArgumentParser()
    parser.add_argument("image", type=str, help='complete path of image including its name.')
    parser.add_argument("outputpath", type=str, help='output path')
    parser.add_argument("scalingFactor", type=float, help='scaling factor of original image')
    parser.add_argument("--charOffset", type = int, default=0, help='number of extra blank spaces')
    parser.add_argument("--white_bg", type=bool, default=0, help='true/false')
    parser.add_argument("--bw_font", type=bool, default=0, help='true/false')
    parser.add_argument("--fontScale", type=int, default=1)
    parser.add_argument("--thickness", type=int, default=1)
    parser.add_argument("--lineType", type=int, default=1)
    parser.add_argument("--textOffset", type=float, default=1.3, help='space btw two characters')
    args = parser.parse_args()


    st = monotonic()
    img = image2ascii(cv2.imread(args.image), args.scalingFactor, args.charOffset, args.white_bg, args.bw_font, args.fontScale, args.thickness, args.lineType, args.textOffset)
    cv2.imwrite(args.outputpath, img)
    et = monotonic()
    print(f"Time Taken : {et-st}sec.")

File: image2ascii.cpp
#include <iostream>
#include <time.h>
#include <opencv2/opencv.hpp>
#include <vector>
#include <algorithm>
#include <string>

using namespace std;

void showHelp()
{
    cout << "Usage: ./ascii_art <image_path> [options]\n\n"
         << "Options:\n"
         << "  -h, --help           Show this help message and exit\n"
         << "  -out, --output <str> Output image path (default: output_ascii.png)\n"
         << "  -s <float>           Scaling factor to resize the input image (default: 0.5)\n"
         << "  -o <int>             Character offset (default: 0)\n"
         << "  -w                   Use a white background (default: black background)\n"
         << "  -b                   Use black/white font color instead of image colors (default: colored)\n"
         << "  -f <int>             OpenCV Font face (default: 0 / FONT_HERSHEY_SIMPLEX)\n"
         << "  --font-scale <float> Scale of the font (default: 0.4)\n"
         << "  -t <int>             Thickness of the font lines (default: 1)\n"
         << "  -l <int>             Line type (default: 1 / LINE_AA)\n"
         << "  --text-offset <float> Padding/offset multiplier for text positioning (default: 1.3)\n"
         << "\nExample:\n"
         << "  ./ascii_art input.jpg -out result.png -s 0.2 -w -b --font-scale 0.5 -t 2\n";
}

char getChar(int color, int offset = 0, bool white_bg = false)
{
    string chars = " .',`:_;-!liI^rv1/ftj~><L*J7T+y?)(cnus=xYVzF}{oha][kAeC4wUX3bdpqZP2EH05GSgOK96DmNR8QBWM&%#@$";
    chars.insert(0, offset, ' ');
    if (white_bg) {
        reverse(chars.begin(), chars.end());
    }
   
    return chars[(color * chars.length()) / 256];
}

cv::Mat image2ascii(cv::Mat image, float scaling_factor, int char_offset = 0, bool white_bg = false, bool bw_font = false, cv::HersheyFonts font_face = cv::FONT_HERSHEY_SIMPLEX, float font_scale = 0.4, int thickness = 1, int line_type = 1, float text_offset = 1.3)
{
    cv::resize(image, image, cv::Size(), scaling_factor, scaling_factor);

    int height = image.rows;
    int width = image.cols;

    cv::Mat gray_image;
    cv::cvtColor(image, gray_image, cv::COLOR_BGR2GRAY);

    int baseline = 0;
    cv::Size textSize = cv::getTextSize("A",
                                        font_face,
                                        font_scale, thickness, &baseline);

    int font_padding = textSize.height * text_offset;

    int ascii_height = height * font_padding;
    int ascii_width = width * font_padding;

    cv::Mat ascii_image;
    if (white_bg) {
        ascii_image = cv::Mat(ascii_height, ascii_width, CV_8UC3, cv::Scalar(255, 255, 255));
    } else {
        ascii_image = cv::Mat(ascii_height, ascii_width, CV_8UC3, cv::Scalar(0, 0, 0));
    }

    if (bw_font) {
        cv::Scalar textColor = white_bg ? cv::Scalar(0, 0, 0) : cv::Scalar(255, 255, 255);
        
        for (int y = 0; y < height; y++)
        {
            for (int x = 0; x < width; x++)
            {
                char text = getChar(gray_image.at<uchar>(y, x), char_offset, white_bg);
                cv::putText(ascii_image,
                            string(1, text),
                            cv::Point(x * font_padding, y * font_padding),
                            font_face,
                            font_scale, textColor, thickness, line_type);
            }
        }
    } else {
        for (int y = 0; y < height; y++)
        {
            for (int x = 0; x < width; x++)
            {
                char text = getChar(gray_image.at<uchar>(y, x), char_offset, white_bg);
                cv::Vec3b pixel = image.at<cv::Vec3b>(y, x);
                cv::Scalar color(pixel[0], pixel[1], pixel[2]);
                cv::putText(ascii_image,
                            string(1, text),
                            cv::Point(x * font_padding, y * font_padding),
                            font_face,
                            font_scale, color, thickness, line_type);
            }
        }
    }

    return ascii_image;
}

int main(int argc, char** argv)
{
    // If no arguments passed, show help
    if (argc < 2) {
        showHelp();
        return 1;
    }

    // Default parameters
    string image_path = "";
    string output_path = "output_ascii.png"; // Added default output path
    float scaling_factor = 0.5f; 
    int char_offset = 0;
    bool white_bg = false;
    bool bw_font = false;
    int font_face = cv::FONT_HERSHEY_SIMPLEX;
    float font_scale = 0.4f;
    int thickness = 1;
    int line_type = 1; 
    float text_offset = 1.3f;

    // Parse command-line arguments
    for (int i = 1; i < argc; ++i) {
        string arg = argv[i];
        
        if (arg == "-h" || arg == "--help") {
            showHelp();
            return 0;
        } else if ((arg == "-out" || arg == "--output") && i + 1 < argc) { // Added parsing for output path
            output_path = argv[++i];
        } else if (arg == "-s" && i + 1 < argc) {
            scaling_factor = atof(argv[++i]);
        } else if (arg == "-o" && i + 1 < argc) {
            char_offset = atoi(argv[++i]);
        } else if (arg == "-w") {
            white_bg = true;
        } else if (arg == "-b") {
            bw_font = true;
        } else if (arg == "-f" && i + 1 < argc) {
            font_face = atoi(argv[++i]);
        } else if (arg == "--font-scale" && i + 1 < argc) {
            font_scale = atof(argv[++i]);
        } else if (arg == "-t" && i + 1 < argc) {
            thickness = atoi(argv[++i]);
        } else if (arg == "-l" && i + 1 < argc) {
            line_type = atoi(argv[++i]);
        } else if (arg == "--text-offset" && i + 1 < argc) {
            text_offset = atof(argv[++i]);
        } else {
            // Assume the argument is the image path if it's not a known flag
            image_path = arg;
        }
    }

    if (image_path.empty()) {
        cerr << "Error: No input image provided!" << endl;
        showHelp();
        return 1;
    }

    clock_t start = clock();
    
    // Load the image
    cv::Mat img = cv::imread(image_path);
    if (img.empty()) {
        cerr << "Error: Could not read image from " << image_path << endl;
        return 1;
    }
    
    cout << "Generating ASCII art... Please wait." << endl;
    
    // Process the image
    cv::Mat result = image2ascii(
        img, 
        scaling_factor, 
        char_offset, 
        white_bg, 
        bw_font, 
        (cv::HersheyFonts)font_face, 
        font_scale, 
        thickness, 
        line_type, 
        text_offset
    );
    
    // Save using the custom output path
    cv::imwrite(output_path, result);
    cout << "Success! Saved output to: " << output_path << endl;
    
    clock_t end = clock();
    double time_taken = (double) (end - start) * 1000 / CLOCKS_PER_SEC;
    cout << "Time Taken: " << time_taken << "ms" << endl;

    return 0;
}